src/libcore/array/mod.rs

```diff
@@ -202,6 +204,65 @@ where
     }
 }
 
+#[stable(feature = "array_from_iter_impl", since = "1.44.0")]
+impl<T, const N: usize> FromIterator<T> for [T; N] 
+where
+     [T; N]: LengthAtMost32,
+{
+    #[inline]
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        <Self as ArrayFromIter<T, I::IntoIter>>::from_iter(iter.into_iter())
+    }
+}
+
+// Specialization trait used for array::from_iter
+trait ArrayFromIter<T, I> {
+    fn from_iter(iter: I) -> Self;
+}
+
+impl<T, I, const N: usize> ArrayFromIter<T, I> for [T; N] 
+    where I: Iterator<Item=T>,
+    [T; N]: LengthAtMost32,
+{
+    default fn from_iter(mut iter: I) -> Self {
+        let mut array: [MaybeUninit<T>; N] = MaybeUninit::uninit_array();
+        
+        for p in array.iter_mut() {
+            p.write(iter.next().expect("Iterator is to short"));
```

Comment:
On panic, you have a potential for a memory leak here. An `OnDrop` functionality should be used to drop the hitherto accumulated array to avoid the leak. (Please also add tests to ensure there is no leak, and more tests in general.)